### PR TITLE
Add more compact serializers for non human readable serialization

### PR DESCRIPTION
**Pull request by Manishearth**
This does not change how TinyStrAuto serializes because there would be a bunch of overhead in making it work there. I tried, and realized it would basically be slower depending on how I use the serialization infrastructure, for not much gain.

### CODE REVIEW

File: src/macros.rs [hunk 1]
```diff
@@ -77,6 +77,52 @@ macro_rules! tinystr16 {
...
+                use std::string::ToString;
+
+                if deserializer.is_human_readable() {
+                    let x: Cow<'de, str> = serde::Deserialize::deserialize(deserializer)?;
+                    x.parse().map_err(|e: Error| SerdeError::custom(e.to_string()))
+                } else {
+                    // little-endian
+                    let le = serde::Deserialize::deserialize(deserializer)?;
+                    let bytes = $int::from_le(le).to_ne_bytes();
+                    if let Err(e) = std::str::from_utf8(&bytes) {
```
Comment (zbraniecki) [inline src/macros.rs:114]: Does that basically parse the bytes to string, and then line below parse it to TinyStr?

It seems to me that we either trust the source and can skip that, or we don't trust the source and then we should use `TinyStr::from_bytes`, no?

Comment (Manishearth) [reply src/macros.rs:114]: We can yeah, I was hoping to avoid the extra length check

File: src/macros.rs [hunk 2]
```diff
@@ -77,6 +77,52 @@ macro_rules! tinystr16 {
...
+                        return Err(SerdeError::custom(e.to_string()))
+                    }
+                    unsafe {
+                        Ok(Self::new_unchecked(le))
+                    }
```
Comment (sffc) [inline src/macros.rs:119]: I strongly object to having an `unsafe` hidden inside a deserializer.  Unless we can mark the whole deserializer implementation as `unsafe`, then we need to do more work to verify that the byte sequence is valid for a TinyStr.

In other words, just use `TinyStr::from_bytes` to both validate and construct.

File: tests/serde.rs [hunk 3]
```diff
@@ -11,6 +11,10 @@ macro_rules! test_roundtrip {
             assert_eq!(json_string, expected_json);
             let recover: $ty = serde_json::from_str(&json_string).unwrap();
             assert_eq!(&*tiny, &*recover);
+
+            let bin = bincode::serialize(&tiny).unwrap();
```
Comment (sffc) [inline tests/serde.rs:15]: Consider putting the bincode bytes themselves into the test file like you did with JSON.

**Review by sffc — CHANGES_REQUESTED** [review]

**Review by zbraniecki — APPROVED** [review]

File: tests/serde.rs (hunk 3 above)
Comment (Manishearth) [reply tests/serde.rs:15]: Done!

**Review by zbraniecki — APPROVED** [review]

**Review by sffc — APPROVED** [review]
I'm satisfied:

```
sffc@sffc:~/projects/tinystr$ cross test --target powerpc-unknown-linux-gnu --all-features test_roundtrip
     Running /target/powerpc-unknown-linux-gnu/debug/deps/serde-8eb225fc36fe899e

running 6 tests
test test_roundtrip16 ... ok
test test_roundtrip4_1 ... ok
test test_roundtrip4_2 ... ok
test test_roundtrip8 ... ok
test test_roundtripauto_1 ... ok
test test_roundtripauto_2 ... ok

test result: ok. 6 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.03s
```

**Comment by Manishearth** [comment]
Posting our endianness doc for posterity

image

**Pull request merged**